### PR TITLE
Add Bitcoin Gold implementation

### PR DESCRIPTION
**Pull request by h4x3rotab**
## Description

(Update since #911: fixed android and ios integration tests)

Bitcoin Gold is already added to the [trustwallet/assets](https://github.com/trustwallet/assets/pull/622). This PR adds the implementation including the address codec and transaction signing logic.

## Testing instructions

Passes unit tests: `./build/tests/tests tests`
Create a new address in walletconsole:

```
> coin btg
Set active coin to: bitcoingold    Use 'coin' to change.  (name: 'bitcoin gold'  symbol: btg  numericalid: 156)
> addrDefault
Result:  btg1q8hnwn2j2vges5u9h9t456zfd46sjxfmxvk3rvl
```

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description (e.g. 'Fixes #444 ).

### CODE REVIEW

**Comment by h4x3rotab** [comment]
https://dev.azure.com/TrustWallet/Trust%20Wallet%20Core/_build/results?buildId=26514&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=40b1ee41-44d6-5bba-aa04-4b76a5c732e5&l=14

> docker: Error response from daemon: pull access denied for trustwallet/wallet-core, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.

Maybe flaky CI errors?

**Comment by hewigovens** [comment]
Hi, We have no plan to add BTG (or other bitcoin forks) to Trust Wallet

**Comment by catenocrypt** [comment]
The iOS CI build failed with this error:

```
Testing failed:
	Type 'CoinType' has no member 'bitcoingold'
	Testing cancelled because the build failed.
```

Maybe BitcoinGold?

**Comment by catenocrypt** [comment]
The Linux CI build failure is infrastructure-related, we will check.

**Comment by h4x3rotab** [comment]
> Hi, We have no plan to add BTG (or other bitcoin forks) to Trust Wallet

Hi Hewig, yeah, we fully understood that. We checked the integration guidance and our first thought was that merging it into wallet-core could benefit more projects and the open source blockchain space, even given TrustWallet is not going to add it. Btw would you mind introduce us the go-to person for a chat? We would really appreciate it.

File: src/BitcoinGold/Address.cpp [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/Address.cpp:1]: remove copy/paste code, I don't think we need this

File: src/BitcoinGold/Address.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/Address.h:1]: remove copy/paste code, I don't think we need this

File: src/BitcoinGold/SegwitAddress.cpp [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/SegwitAddress.cpp:1]: remove copy/paste code, I don't think we need this

File: src/BitcoinGold/SegwitAddress.h [hunk 4]
```diff
@@ -0,0 +1,45 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/SegwitAddress.h:1]: remove copy/paste code, I don't think we need this

File: src/BitcoinGold/Signer.cpp [hunk 5]
```diff
@@ -0,0 +1,12 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/Signer.cpp:1]: remove copy/paste code, I don't think we need this

File: src/BitcoinGold/Signer.h [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/BitcoinGold/Signer.h:1]: remove copy/paste code, I don't think we need this

File: src/proto/BitcoinGold.proto [hunk 7]
```diff
@@ -0,0 +1,11 @@
+// Copyright © 2017-2020 Trust Wallet.
```
Comment (hewigovens) [inline src/proto/BitcoinGold.proto:1]: remove copy/paste code, I don't think we need this

File: swift/Tests/CoinAddressDerivationTests.swift [hunk 8]
```diff
@@ -190,8 +193,6 @@ class CoinAddressDerivationTests: XCTestCase {
                 case .filecoin:
                     let expectedResult = "f1zzykebxldfcakj5wdb5n3n7priul522fnmjzori"
                     AssetCoinDerivation(coin, expectedResult, derivedAddress, address)
-                @unknown default:
```
Comment (hewigovens) [inline swift/Tests/CoinAddressDerivationTests.swift:193]: revert please

File: tests/BitcoinGold/TWAddressTests.cpp [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+    ASSERT_TRUE(BitcoinGold::Address::isValid(std::string(ADDRESS)));
+    ASSERT_FALSE(BitcoinGold::Address::isValid(std::string(FAKEADDRESS)));
+}
+
+TEST(TWBitcoinGoldAddress, PubkeyToAddress) {
+    const auto privateKey = PrivateKey(parse_hex(PRIVATE_KEY));
+    const auto publicKey = privateKey.getPublicKey(TWPublicKeyTypeSECP256k1);
+
+    /// construct with public key
+    auto address = BitcoinGold::Address(PublicKey(publicKey), 38);
```
Comment (hewigovens) [inline tests/BitcoinGold/TWAddressTests.cpp:32]: replace magic number 38 with `p2pkhPrefix` in `Coin.h`

File: tests/BitcoinGold/TWSignerTests.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+using namespace TW;
+using namespace TW::Bitcoin;
+
+TEST(TWBitcoinGoldSigner, SignTransaction) {
+    const int64_t amount = 10000;
+
+    // Setup input
+    Proto::SigningInput input;
+    input.set_coin_type(TWCoinTypeBitcoinGold);
+    input.set_hash_type(TWBitcoinSigHashTypeFork | TWBitcoinSigHashTypeAll | (79 << 8));
```
Comment (hewigovens) [inline tests/BitcoinGold/TWSignerTests.cpp:31]: Add `79 << 8` to `TWBitcoinSigHashType`?

File: tests/BitcoinGold/TWSignerTests.cpp [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    // Sign
+    auto txSinger = TransactionSigner<Transaction, TransactionBuilder>(std::move(input));
+    txSinger.transaction.lockTime = 0x00098971;
+    auto result = txSinger.sign();
+
+    ASSERT_TRUE(result) << result.error();;
+    auto signedTx = result.payload();
+
+    Data serialized;
+    signedTx.encode(true, serialized);
```
Comment (hewigovens) [inline tests/BitcoinGold/TWSignerTests.cpp:69]: can you broadcast a tx signed by wallet core and paste the tx id here?

File: tests/BitcoinGold/TWSignerTests.cpp (hunk 10 above)
Comment (tolak) [reply tests/BitcoinGold/TWSignerTests.cpp:31]: Yes, 79 is the forkid of Bitcoin Gold. So here we set FORKID bits of TWBitcoinSigHashType to identify Bitcoin Gold signature.

File: tests/BitcoinGold/TWSignerTests.cpp (hunk 11 above)
Comment (tolak) [reply tests/BitcoinGold/TWSignerTests.cpp:69]: Hi, we already added lockscrip&zpub test in latest commit:11aa44f and had broadcast a tx signed by wallet core, mainnet txid: 3e818ad25d73123b6c1c8099ed462aa5413a4ef57d66d9d260306c012753ba43. You can explore it here: https://btg2.trezor.io/

File: src/proto/BitcoinGold.proto (hunk 7 above)
Comment (catenocrypt) [reply src/proto/BitcoinGold.proto:1]: This proto file is not needed

Comment (tolak) [reply src/proto/BitcoinGold.proto]: removed!

**Comment by h4x3rotab** [comment]
> Missing lock script / zpub tests:
> 
> https://developer.trustwallet.com/wallet-core/newblockchain#bitcoin-forks-checklist

Added at 11aa44f. Please take a look :)

File: tests/BitcoinGold/TWSignerTests.cpp (hunk 11 above)
Comment (hewigovens) [reply tests/BitcoinGold/TWSignerTests.cpp:69]: can you add the link as comment in code?

**Comment by hewigovens** [comment]
please rebase master, Android CI should be fixed

File: .github/CODEOWNERS [hunk 12]
```diff
@@ -0,0 +1,6 @@
+# These owners will be the default owners for everything in
```
Comment (hewigovens) [inline .github/CODEOWNERS:1]: can you rebase (not merge) and create a clean pr? those files should not be included in this pr

Comment (tolak) [reply .github/CODEOWNERS]: I rebased those again, it's a clean pr now, FYI.

File: tests/BitcoinGold/TWSignerTests.cpp (hunk 11 above)
Comment (tolak) [reply tests/BitcoinGold/TWSignerTests.cpp:69]: Already added the link as comment in code at the latest commit: 01e56ab. FYI

File: .github/CODEOWNERS (hunk 12 above)
Comment (tolak) [reply .github/CODEOWNERS]: Sorry we lost some file during this rebase, we will handle it as soon.

Comment (catenocrypt) [reply .github/CODEOWNERS]: The PR contains only 3 new files now (2 tests), I think rebase was not right.

Comment (tolak) [reply .github/CODEOWNERS]: Yes, something wrong during this rebase, i am fixing it.

Comment (tolak) [reply .github/CODEOWNERS]: Already fixed!

File: tests/BitcoinGold/TWSignerTests.cpp (hunk 11 above)
Comment (hewigovens) [reply tests/BitcoinGold/TWSignerTests.cpp:69]: yes, I can see in https://github.com/trustwallet/wallet-core/commit/01e56ab97876d02f3b38de6cb8a5b77e5bd3649a but missing here: https://github.com/trustwallet/wallet-core/blob/6d02ae2d89d3877f309a14317067d6f68c0c5abc/tests/BitcoinGold/TWBitcoinGoldTests.cpp#L66

File: tests/BitcoinGold/TWSignerTests.cpp [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    ASSERT_EQ(hex(serialized),
```
Comment (hewigovens) [inline tests/BitcoinGold/TWSignerTests.cpp:70]:
```suggestion
    // BitcoinGold Mainnet: https://btg2.trezor.io/tx/3e818ad25d73123b6c1c8099ed462aa5413a4ef57d66d9d260306c012753ba43
    ASSERT_EQ(hex(serialized),
```

**Comment by h4x3rotab** [comment]
Looks like we are ready to go :)
